python/cuxfilter/charts/datashader/custom_extensions/holoviews_datashader.py

```diff
@@ -0,0 +1,577 @@
+            tile_sources[self.tile_provider]()
+            if (self.tile_provider is not None)
+            else self.tile_provider
+        )
+        self.nodes_chart = InteractiveDatashaderPoints(
+            source_df=self.nodes_df,
+            x=self.node_x,
+            y=self.node_y,
+            aggregate_col=self.node_aggregate_col,
+            aggregate_fn=self.node_aggregate_fn,
+            color_palette=self.node_color_palette,
+            pixel_shade_type=self.node_pixel_shade_type,
+            tile_provider=self.tile_provider,
+            legend=self.legend,
+            legend_position=self.legend_position,
+            spread_threshold=self.node_spread_threshold,
+            point_shape=self.node_point_shape,
+            max_px=self.node_max_px,
+        )
+
+        self.edges_chart = InteractiveDatashaderLine(
+            source_df=self.edges_df,
+            x=self.edge_source,
+            y=self.edge_target,
+            color=self.edge_color,
+            transparency=self.edge_transparency,
+        )
+
+    def update_data(self, nodes=None, edges=None):
+        if nodes:
+            self.nodes_chart.update_data(nodes)
+        if edges:
+            self.edges_chart.update_data(edges)
+
+    def view(self):
+        def set_tools(plot, element):
+            if plot.state.toolbar.tools[-1] != self.display_edges:
+                plot.state.add_tools(self.inspect_neighbors)
+                plot.state.add_tools(self.display_edges)
+
+        dmap_nodes = dynspread(
+            self.nodes_chart.get_chart(
+                streams=[
+                    self.box_stream,
+                    self.lasso_stream,
+                    self.reset_stream,
+                ]
+            ),
+            threshold=self.node_spread_threshold,
+            shape=self.node_point_shape,
+            max_px=self.node_max_px,
+        ).opts(
+            xaxis=None,
+            yaxis=None,
+            responsive=True,
+            default_tools=[],
+            active_tools=["wheel_zoom", "pan"],
+            tools=self.tools,
+            hooks=[set_tools],
+        )
+
+        dmap_edges = dynspread(
+            self.edges_chart.get_chart().opts(default_tools=[])
+        )
+
+        dmap_graph = dmap_edges * dmap_nodes
+
+        return pn.pane.HoloViews(
+            self.tiles * dmap_graph if self.tiles is not None else dmap_graph,
+            sizing_mode="stretch_both",
+            height=self.height,
+        )
+
+    def add_box_select_callback(self, callback_fn):
+        self.box_stream = hv.streams.SelectionXY(subscribers=[callback_fn])
+
+    def add_lasso_select_callback(self, callback_fn):
+        self.lasso_stream = hv.streams.Lasso(subscribers=[callback_fn])
+
+    def reset_all_selections(self):
+        self.lasso_stream.reset()
+        self.box_stream.reset()
+
+    def add_reset_event(self, callback_fn):
+        self.reset_stream = hv.streams.PlotReset(subscribers=[callback_fn])
```

Comment:
There appears to a fair amount of overlap with `InteractiveDatashader` (these callbacks, several params). Maybe there is a further base class that can be factored out of `InteractiveDatashader`, or perhaps a small bit of functionality can be moved into its subclasses so that it can also be used as a base class here?

Reply:
ah I see what you mean, yes that's certainly possible to create another baseclass and add `InteractiveDatashader` as its subclass

Reply:
added `InteractiveDatashaderBase` class with all the overlapping elements